libraries/deprecations.rb

```diff
@@ -0,0 +1,62 @@
+# encoding: utf-8
+# This module is use to support existing array as well as hash of hashes format
+# that are being assigned by end user.
+#
+# In order to provide backward compatibility this module extended by profiles key object
+#
+module Deprecations
+  module Audit
+    # ArrayProfile module extend the hash instance object node['audit']['profiles']
+    # to add warning & format the array assignment.
+    module ArrayProfile
+      # @overload []=(*args)
+      # This method provide the support to assign item as hash(key as String|Symbol)
+      #
+      # @param [String|Symbol] key assigned add as name.
+      # @param [Hash] profile attributes the are being assigned.
+      #
+      # @example format the profile assignment if it is hash of hashes format.
+      #
+      #    node.default['audit']['profiles']['linux'] = { 'compalince': 'base/linux' }
+      #
+      # @example Formatted response return
+      #    { name: linux, 'compalince': 'base/linux' }
+      #
+      # @return [Array<Profiles>]
+      def []=(*args)
+        self << { name: args.first }.merge(Hash.try_convert(args.last) || {})
+      end
+
+      # @overload push(*args)
+      # WARN msg for the use of array assigment.
+      # @param [Array] Profiles the are being assigned.
+      #
+      # @return [Array<Profiles>]
+      def push(*args)
```

Comment:
There are more Array methods that are missing, such as `<<` that concern us that this will be buggy even if it is temporary.

Reply:
Yeah so the fact that this is being widely used and has a large blast radius to every single user, while it makes going the extra mile to make backcompat work also means that you have to assume that people will be using entirely arbitrary array mutation methods on this attribute.

We do have a complete list of those in Chef::Node::Mixin::ImmutablizeArray::DISALLOWED_MUTATOR_METHODS and a complete list of Hash mutators in Chef::Node::Mixin::ImmutablizeHash::DISALLOWED_MUTATOR_METHODS.

The ones that are specific to Arrays and not Hashes could be included here.  What would we do about the ones that are common across Hashes and Arrays though?  Can't determine if the user intends to be coming from Hash-space or Array-space in that case.  I suppose we could always assume that we need backcompat first and use Array space until the compat layer is removed.

Reply:
Added warning for all  `Array` mutator methods.

For the `Hash` and common mutator methods does not do anything because changing the Hash methods behaviour leads to confusion after the array of hashes support will be deprecated.

For example if we provide `.merge` method support to profiles it return  `Array of hashes` instead of `Hash of hashes`.

Please let me know your thoughts?